File: src/2707. Extra Characters in a String.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
function minExtraChar(s, dictionary) {
  const dp = Array(s.length + 1).fill(0)
  for (let i = 1; i <= s.length; ++i) {
    dp[i] = dp[i - 1] + 1
    for (let c of dictionary) {
      if (i - c.length >= 0 && s.substring(i - c.length, i) === c) {
        dp[i] = Math.min(dp[i], dp[i - c.length])
      }
    }
  }
  return dp[dp.length - 1]
}

File: src/2707. Extra Characters in a String.py
class Solution:
    def minExtraChar(self, s: str, dictionary: List[str]) -> int:
        dp = [0] * (len(s) + 1)
        for i in range(1, len(s) + 1):
            dp[i] = dp[i - 1] + 1
            for c in dictionary:
                if i - len(c) >= 0 and s[i - len(c):i] == c:
                    dp[i] = min(dp[i], dp[i - len(c)])
        return dp[-1]